### PR TITLE
libimagstore: Eliminate EntryHeader type

### PR DESCRIPTION
**Pull request by matthiasbeyer**
Followup for #835 

This eliminates the `EntryHeader` type by moving the functionality to traits implemented on `toml::Value`.

Refactoring is about to be done.

### CODE REVIEW

**Comment by matthiasbeyer** [comment]
Not sure why the binary tests fail. @mario-kr would you like to test? I will continue fixing the library tests, but would be awesome if you could test the binary tests.

**Comment by matthiasbeyer** [comment]
So as it seems, the headers are screwed up somehow. `imag-store create` does not create an entry anymore, but the debug output does not provide a error message, so I guess there is some serious bug in here.

---

One thing that should be done before merging this is moving the free functions out of the public scope here. We introduce the public module `toml_ext` from the store, which contains public utility functions (because we need them in `store` and `toml_ext`). We should move them to a `util` module.

**Comment by matthiasbeyer** [comment]
@TheNeikos The above failed - but when we tested it, it didn't. That's because travis here uses rustc 1.11 still, and we are on 1.14 ... hmmmm

**Comment by mario-kr** [comment]
rebasing the branch might help considerung the `.travis.yml`.

**Comment by matthiasbeyer** [comment]
@mario-kr that'd be a fix, yes. Anyways, I refactored the thing to use a custom macro internally, which should also work on rustc 1.11!

**Comment by matthiasbeyer** [comment]
Rebased. @mario-kr was probably right ... still, I included the custom macro patch, so we either define `if_cfg_panic!()` as empty or as wrapper to `panic!()`. Could be of use later, too.

**Comment by mario-kr** [comment]
Ah, the regex update is throwing sticks between your feet.

**Comment by matthiasbeyer** [comment]
Jup. That would have caused a CI-failure after merge, so good that I rebased this!

**Comment by matthiasbeyer** [comment]
Now things get funny! :smile:

**Comment by matthiasbeyer** [comment]
@mario-kr rebased, fixupd, Ci succeeds... what do we want more? Feel free to review now! :smile:

File: libimagentrylink/src/external.rs [hunk 1]
```diff
@@ -312,6 +313,8 @@ impl ExternalLinker for Entry {
 
     /// Set the external links for the implementor object
     fn set_external_links(&mut self, store: &Store, links: Vec<Url>) -> Result<()> {
+        use libimagstore::toml_ext::TomlValueExt;
```
Comment (mario-kr) [inline libimagentrylink/src/external.rs]: is this not included at the top of the file already?

File: libimagentrytag/src/tagable.rs [hunk 2]
```diff
@@ -76,6 +77,8 @@ impl Tagable for EntryHeader {
     }
 
     fn set_tags(&mut self, ts: &[Tag]) -> Result<()> {
+        use libimagstore::toml_ext::TomlValueExt;
```
Comment (mario-kr) [inline libimagentrytag/src/tagable.rs]: same here? Am i missing a detail of rust here?

File: libimagref/src/reference.rs [hunk 3]
```diff
@@ -390,6 +391,8 @@ impl<'a> Ref<'a> {
     /// Update the Ref by re-checking the file from FS using the passed Hasher instance
     /// This errors if the file is not present or cannot be read()
     pub fn update_ref_with_hasher<H: Hasher>(&mut self, h: &H) -> Result<()> {
+        use libimagstore::toml_ext::TomlValueExt;
```
Comment (mario-kr) [inline libimagref/src/reference.rs]: same

File: libimagstore/Cargo.toml [hunk 4]
```diff
@@ -39,3 +39,8 @@ env_logger = "0.3"
 default = []
 verify  = []
 
+# Enable panic!()s if critical errors occur.
+# Can be used to debug the store more intensly via the imag-store commandline
+# application
```
Comment (mario-kr) [inline libimagstore/Cargo.toml]: an example would be nice.

**Comment by matthiasbeyer** [comment]
I should probably rewrite the history of this branch a bit more, so we can have a nice history.

**Comment by matthiasbeyer** [comment]
Now I consider this history in a mergeable shape! :small_airplane:

File: libimagstore/src/store.rs [hunk 5]
```diff
@@ -1147,11 +984,26 @@ impl Entry {
     pub fn new(loc: StoreId) -> Entry {
         Entry {
             location: loc,
-            header: EntryHeader::new(),
+            header: Entry::default_header(),
             content: EntryContent::new()
         }
     }
 
+    pub fn default_header() -> Value { // BTreeMap<String, Value>
```
Comment (mario-kr) [inline libimagstore/src/store.rs]: further down (in `toml_ext` somewhere), you have a trait called `Header`, which has s function with the same name.
Is it possible, to also implement this trait for `Entry`? If not, could you make a call here to `Value::default_header()` to reduce duplicate code? It seems to do the exact same thing...

File: libimagstore/src/toml_ext.rs [hunk 6]
```diff
@@ -353,10 +358,104 @@ impl Extract for Value {
...
+        use toml::Parser;
+
+        let mut parser = Parser::new(s);
+        parser.parse()
+            .ok_or(ParserErrorKind::TOMLParserErrors.into())
+            .and_then(verify_header_consistency)
+            .map(Value::Table)
+    }
+
+    fn default_header() -> Value {
```
Comment (mario-kr) [inline libimagstore/src/toml_ext.rs]: see previous comment

File: libimagstore/src/util.rs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+macro_rules! if_cfg_panic {
+    ()                       => { panic!() };
+    ($msg:expr)              => { panic!($msg) };
+    ($fmt:expr, $($arg:tt)+) => { panic!($fmt, $($($arg),*)) };
+}
+
+#[cfg(not(feature = "early-panic"))]
+#[macro_export]
+macro_rules! if_cfg_panic {
+    ()                       => { };
```
Comment (mario-kr) [inline libimagstore/src/util.rs]: is it not possbile to do just sth like `($i)* => {};`?

Comment (matthiasbeyer) [inline libimagstore/src/util.rs]: I used the macro implementation of `panic!()` itself as reference here, so ... no, I won't change that!

Comment (mario-kr) [inline libimagstore/src/util.rs]: ah, ok. was just wondering if there'd be an easier way.

File: libimagstore/src/store.rs (hunk 5 above)
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Fixed up with 337a03b.

It is not possible to implement `Header` for `Entry`, as an Entry is not a Header. The `Header` trait is _only_ for extending the `toml::Value` type. But I fixed what you suggested by calling `Value::default_header()` in this function.

**Comment by matthiasbeyer** [comment]
Oh yeah... food would be a nice thing now! Lets see whether CI says yes, then rebase (fixup) and then I'll grab some food to (procrastinating from learning for my exams... shame on me)!

**Pull request merged**